js_modules/dagster-ui/packages/ui-core/src/app/Route.tsx

```diff
@@ -0,0 +1,76 @@
+import {ComponentProps, ReactNode, memo, useMemo} from 'react';
+import {Route as ReactRouterRoute, useRouteMatch} from 'react-router-dom';
+import {useSetRecoilState} from 'recoil';
+
+import {currentPageAtom} from './analytics';
+import {useDangerousRenderEffect} from '../hooks/useDangerousRenderEffect';
+
+export const Route = memo((props: ComponentProps<typeof ReactRouterRoute>) => {
+  const {render, children, component: Component} = props;
+
+  const renderFn = useMemo(() => {
+    if (!render) {
+      return;
+    }
+    return (...args: Parameters<typeof render>) => {
+      return (
+        <>
+          <Wrapper>{render(...args)}</Wrapper>
+        </>
+      );
+    };
+  }, [render]);
+  const WrapperComponent = useMemo(() => {
+    if (!Component) {
+      return;
+    }
+    return (props: any) => {
+      return (
+        <>
+          <Wrapper>
+            <Component {...props} />
+          </Wrapper>
+        </>
+      );
+    };
+  }, [Component]);
+
+  const childRenderFn = useMemo(() => {
+    if (!(children instanceof Function)) {
+      return;
+    }
+    return (...args: Parameters<typeof children>) => {
+      return (
+        <>
+          <Wrapper>{children(...args)}</Wrapper>
+        </>
+      );
+    };
+  }, [children]);
+
+  if (render) {
+    return <ReactRouterRoute {...props} render={renderFn} />;
+  }
+  if (Component) {
+    return <ReactRouterRoute {...props} component={WrapperComponent} />;
+  }
+  if (children instanceof Function) {
+    return <ReactRouterRoute {...props}>{childRenderFn}</ReactRouterRoute>;
+  }
+  return (
+    <ReactRouterRoute {...props}>
+      <Wrapper>{children}</Wrapper>
+    </ReactRouterRoute>
+  );
+});
+
+const Wrapper = memo(({children}: {children: ReactNode}) => {
+  const {path} = useRouteMatch();
+
+  const setCurrentPage = useSetRecoilState(currentPageAtom);
+  useDangerousRenderEffect(() => {
+    setCurrentPage(({specificPath}) => ({specificPath, path}));
+  }, [path]);
+
+  return children;
```

Comment:
Ahh I see, so we can't use the standard integration points in react-router's `history` because this wants the `routeMatch` not just the pathname, etc.  Can't think of a better way to do this 👍